docs(cli): add docs for custom plugins (#312)
code-pushup/cli: packages/cli/README.md

      - run: npm ci
      - run: npx code-pushup autorun --upload.apiKey=${{ secrets.PORTAL_API_KEY }}
```

## Custom Plugins

One of the main features of Code PushUp is the ability to write custom plugins to track your own metrics.

> It enables you to implement neraly any kind of metric you want to track with minimum effort.

Reviewer: ```suggestion
It enables you to implement nearly any kind of metric you want to track with minimum effort.
```

In this section we will go through the steps to create a custom plugin and integrate it in your project.

### Set up

> To start crafting custom plugins you need a minimum `code-pushup.config.ts` file including the `persist` and `plugins`

Reviewer 2: ```suggestion
To start crafting custom plugins you need a minimum `code-pushup.config.(ts|js|mjs)` file including the `persist` and `plugins`
```
We support `ts`, `mjs` and `js`, I would either omit the file format or include all options and emphasise the example is for `ts`.

property.

> **config and plugin template**

Reviewer 2: ```suggestion
In the configuration file, the most important properties are:
...

#### Minimum configuration file example
```
It would be great to introduce all the important parts of the configuration file.

```typescript
// code-pushup.config.ts
import { AuditOutputs, PluginConfig } from '@code-pushup/models';

type Options = {};
async function create(options: Options): PluginConfig {
  const audit = {
    slug: 'my-audit',
    title: 'My audit',
    description: 'My custom audit.',
  };
  return {
    slug: 'my-plugin',
    title: 'My plugin',
    icon: 'javascript', // icon name from [vscode-material-icon-theme](https://github.com/PKief/vscode-material-icon-theme/tree/main/icons)
    description: 'My custom plugin.',
    audits: [audit],
    runner: (): AuditOutputs => {
      // Dummy audit output
      const auditOutput: AuditOutput = {
        slug: audit.slug,
        score: 0,
        value: 0,
      };

>       console.log('In plugin runner');

Reviewer 2: ```suggestion
      console.info('In plugin runner');
```
I would be careful with `console.log` as an example of a good practice.

      // return dummy data of type `AuditOutputs` as plugin result
      return [auditOutput];
    },
  };
}

export default {
  persist: {
    outputDir: '.code-pushup',
  },
  plugins: [await create()],
  categories: [],
};
```

Execute the CLI with `npx code-pushup collect --no-progress` and you should the following output:

> **stdout of plugin template**

Reviewer 2: As discussed, it would be valuable to show a realistic output and perhaps mention as a `detail` what the example above would produce.

Additionally, it would be helpful to list what one can expect in the summary.

```sh
Code PushUp Report - @code-pushup/core@x.y.z

My plugin audits
● My audit                                                            0

Categories
┌──────────┬───────┬────────┐
│ Category │ Score │ Audits │
└──────────┴───────┴────────┘

Made with ❤ by code-pushup.dev
```

The CLI argument `--no-progress` is used to get better debugging experience in the console, as the progress bar can
interfere with debugging logs.

> The categories are empty for now. But under the audit listing you can see your plugin title `My plugin` it's listed

Reviewer: ```suggestion
The categories are empty for now. But under the audit listing you can see your plugin title `My plugin`, it's listed
```

audit `My audit` and the resulting value `0`.

### Plugin Runner

The core of a plugin is defined under the `runner` property and can get implemented in 2 ways:

- as a `RunnerFunction`
- as a `RunnerConfig`

> Your preferred way should be the `RunnerFunction` as it is more flexible and easier to use.

Reviewer: This is how I would distinguish these for plugin authors.

```suggestion
We recommend the `RunnerFunction` for getting started, as it's easier to use for simple plugins and can be written in the config file directly.

The `RunnerConfig` is suitable for more complex, performance-heavy plugins (runner executed off the main thread), and is more flexible with regards to runtime (can run any shell command, not restricted to JavaScript).
```

import { AuditOutputs, PluginConfig } from '@code-pushup/models';
import { RunnerFunction } from './plugin-config-runner';

type Options = {
  directory: string;
};

async function create(options: Options): Promise<PluginConfig> {
  const fileSizeAudit = {
    slug: 'file-size-audit',
    title: 'File size audit',
  };
  return {
    slug: 'file-size',
    title: 'File size plugin',
    icon: 'javascript',
    audits: [fileSizeAudit],
    runner: runnerFunction(options),
  };
}

// we use a closure to pass options to the runner function for better DX
async function runnerFunction(options: Options): RunnerFunction {
  return () => {
    // implementation follows in the below section
    return [] as AuditOutputs;
  };
}

export default {
  persist: {
    outputDir: '.code-pushup',
  },
  plugins: [
    await create({
      directory: './src',
    }),
  ],
  categories: [],
};
```

</details>

The first thing we should think about is how to get the raw data to calculate the metrics.
In our case we need to get the file name and size of all files in a directory.

The basic implementation looks like this:

**get file-size data**

```typescript
// code-pushup.config.ts

> // ...

Reviewer: Maybe good to list where the Node APIs are imported from, might not be clear to everyone:

```suggestion
// ...
import { readdir, stat } from 'node:fs/promises';
import { join } from 'node:path';
// ...
```

// get raw file size data
type FileSizeInfo = { file: string; size: number };
async function getFileSizeData(options: Options): Promise<FileSizeInfo[]> {
  const { directory } = options;

  let results = [];
  const files = await readdir(directory);

  for (const file of files) {
    const filePath = join(directory, file);
    const stats = await stat(filePath);

    // if file, get file size info
    if (stats.isFile()) {
      results.push({ filePath, size: stats.size });
    }
    // if directory, recurse
    else if (stats.isDirectory()) {
      results.push(...(await getFileSizeData({ directory: filePath })));
    }
  }

  return results;
}
```

> The above code will recursively get the file size info for all files in the specified directory and it's subdirectories.

Reviewer: ```suggestion
The above code will recursively get the file size info for all files in the specified directory and its subdirectories.
```

Let's create a runner function that uses the above code to get the file size data and turns it into the shape
of `AuditOutputs`:

**file-size `RunnerFunction`**

```typescript
// code-pushup.config.ts
import { pluralize } from '@code-pushup/utils';

// ...

async function runnerFunction(options: Options): Promise<AuditOutputs> {
  const data = await getFileSizeData(options);
  // We use a helper function of the utils package to pluralize the display value
  const displayValue = `${data.length} ${data.length === 1 ? 'file' : pluralize('file')}`;

  let fileSizeAuditOutput: AuditOutput = {
    slug: fileSizeAudit.slug,
    // We have always a score of 1. Proper implementation of scores will be covered in a next section.
    score: 1,
    value: data.length,
    displayValue,
  };

  return [fileSizeAuditOutput];
}
```

Now we can execute the CLI with `npx code-pushup collect --no-progress` and see a similar output as the following:

**stdout of basic file-size plugin**

```sh
Code PushUp Report - @code-pushup/core@x.y.z

File size plugin audits

> ● File size audit                                                            2

Reviewer: At this point I believe it would be `2 files` instead of just `2`, because you've created the `displayValue`:

```suggestion
● File size audit                                                            2 files
```

Categories
┌──────────┬───────┬────────┐
│ Category │ Score │ Audits │
└──────────┴───────┴────────┘

Made with ❤ by code-pushup.dev
```

### RunnerConfig

The second way to write a plugin runner is a `RunnerConfig`.
This option is less flexible but can be used in cases when you have to use another CLI.

> why runner function can't be used...

Reviewer: Perhaps we could have a section (or even separate document) on the pros and cons on RunnerFunction vs RunnerConfig that we can refer to :thinking:

We will implement a performance focused plugin using the [Lighthouse CLI](https://github.com/GoogleChrome/lighthouse#using-the-node-cli) as real life example.

> Let's start with a `crate` function maintaining the basic information of the `PluginConfig`.

Reviewer: ```suggestion
Let's start with a `create` function maintaining the basic information of the `PluginConfig`.
```

Seeing `crate` made me think you'd be implementing it in Rust :grin:

<details>
<summary> <b>lighthouse plugin setup</b> (collapsed for brevity) </summary>

```typescript
// code-pushup.config.ts
import { PluginConfig, RunnerConfig } from '@code-pushup/models';

type Options = {
  url: string;
  onlyAudits: string;
};

async function create(options: Options): Promise<PluginConfig> {
  const lcpAudit = {
    slug: 'largest-contentful-paint',
    title: 'Largest Contnentful Paint',
  };

  return {
    slug: 'lighthouse',
    title: 'Chrome Lighthouse plugin',
    icon: 'lighthouse',
    audits: [
      lcpAudit,
      // other lighthouse audits
    ],
    runner: runnerConfig(options),
  };

> }

Reviewer: Both the `create` function examples return a promise only because of the `async` keyword (no `await` in the body), which necessitates using `await create(...)` in the config. Wouldn't it be simpler to have sync `create` functions instead of having them be artificially async? On the other hand, it's good mention we how to do async create functions somewhere (though perhaps seeing `await eslintPlugin(...)` in our examples is enough of an indication). :thinking:

function runnerConfig(options: Options): RunnerConfig {
  // implementation follows in the below section
  return {} as RunnerConfig;
}

export default {
  // ...
  plugins: [
    await create({
      url: 'https://example.com',
      onlyAudits: 'largest-contentful-paint',
    }),
  ],
  // ...
};
```

</details>

The first thing we should think about is how to get the raw data.

> With the lighthouse CLI it is ease as it already provides a report file in `json` format containing a set of audits.

Reviewer: ```suggestion
With the Lighthouse CLI it is easy as it already provides a report file in `json` format containing a set of audits.
```

The lighthouse CLI can be executed like this: `npx lighthouse https://example.com`  
You should see console output of the audits created by the CLI.

To get better debugging experience we add a couple of more options:

- The format and location of the output can be configured with `--output=json --outputFile=lighthouse-report.json`.  
  This ensures we avoid overwrites of other existing files and is needed to be able to load the generated lighthouse
  report.
- To reduce the output you can execute only specific audits with the `--onlyAudits` option
  e.g.: `--onlyAudits=largest-contentful-paint`.  
  This will significantly reduce the time lighthouse takes to run.

> - If we want to run the script in the background we can execute lighthouse headless with the

Reviewer: ```suggestion
- If we want to run the script in the background we can execute Lighthouse in headless mode with the
```

  flag `--chrome-flags="--headless=new"`.  

>   It also is helpful when executing lighthouse in th CI.

Reviewer: ```suggestion
  It also is helpful when executing Lighthouse in the CI.
```

The basic implementation of a `RunnerConfig` for the above command looks like this:

```typescript
// code-pushup.config.ts
// ...
import { join } from 'path';
import { AuditOutputs } from '@code-pushup/models';

function runnerConfig(options: Options): RunnerConfig {
  const { url, onlyAudits } = options;
  const outputFile = join(process.cwd(), '.code-pushup', 'lighthouse-report.json');
  return {
    // npx lighthouse https://example.com --output=json --outputFile=lighthouse-report.json  --onlyAudits=largest-contentful-paint
    command: 'npx',

>     args: objectToCliArgs({

Reviewer: You should include the import from `@code-pushup/utils`, since this helper hasn't been established before.

      _: ['lighthouse', url],
      output: 'json',
      'output-path': outputFile,
      onlyAudits,
    }),
    outputFile,
    // implementation follows in the next section
    outputTransform,
  };
}

function outputTransform(output: string): AuditOutputs {
  // Return dummy audit outputs.
  // Otherwise the CLI will throw an error as the lighthouse report is not of shape AuditOutputs
  return [
    {
      slug: lcpAudit.slug,

>       score: 0,
>       value: 0,

Reviewer 2: ```suggestion
      score: 78,
      value: 1580,
      displayValue: '1.58 s'
```
I would include some values, so that there isn't a `0` in the `stdout` report but an "actual" result and it is clear what the number/value refers to.

    },
  ] satisfies AuditOutputs;
}
```

Now we can execute the CLI with `npx code-pushup collect --no-progress` and see a similar output as the following:

**stdout of basic lighthouse plugin**

```sh
Code PushUp Report - @code-pushup/core@x.y.z

Chrome Lighthosue audits
● Largest Contentful Paint                                                0

Categories
┌──────────┬───────┬────────┐
│ Category │ Score │ Audits │
└──────────┴───────┴────────┘

Made with ❤ by code-pushup.dev
```

> #### Implement a `outputTransform` function

Reviewer: ```suggestion
#### Implement an `outputTransform` function
```

> As the result of the lighthouse run has a different shape than the required `AuditOutputs` we need to add

Reviewer 2: ```suggestion
As the Lighthouse result has a different shape than the required `AuditOutputs`, we need to add
```

> a `outputTransform` and implement the transform form a lighthouse report to audit outputs.

Reviewer: ```suggestion
an `outputTransform` and implement the transform from a Lighthouse report to audit outputs.
```

**outputTransform for lighthouse report**

```typescript
// code-pushup.config.ts
import { Result } from 'lighthouse';

// ...

function outputTransform(output: string): AuditOutputs {
  // output is content of `lighthouse-report.json` as string so we have to parse it

>   const lhr = JSON.parse(output as Result);

Reviewer: ```suggestion
  const lhr = JSON.parse(output) as Result;
```

  return Object.values(lhr.audits).map(({ id: slug, score, numericValue: value, displayValue, description }) => ({
    slug,
    score: score,
    value: parseInt(value.toString(), 10),
    displayValue,
    description,
  }));
}
```

Test the output by running `npx code-pushup collect --no-progress --format=md`.
The CLI argument `--format=md` will create an additional file containing our created detail information form above.

> You should see a newly created file `report.md` crated in the folder `.code-pushup` in your current working directory.

Reviewer: ```suggestion
You should see a newly created file `report.md` created in the folder `.code-pushup` in your current working directory.
```

Reviewer 2: ```suggestion
You should see a new file `report.md` created in the folder `.code-pushup` in your current working directory.
```
To avoid repeating `created`?

painted. [Learn more about the Largest Contentful Paint metric](https://developer.chrome.com/docs/lighthouse/performance/lighthouse-largest-contentful-paint/)
```

### Scoring of audits

Every audit has a score as floating number between 0 and 1.
We will extend the file-size example to calculate the score based on a budget.

Let's first extend the options object with a `budget` property and use it in the runner config:

**file-size plugin form section [RunnerFunction](#RunnerFunction)**

```typescript
// code-pushup.config.ts
type Options = {
  // ...
  budget: number; // in bytes
};

export default {
  // ...
  plugins: [
    await createPlugin({
      directory: './src',
      budget: 1000,
    }),
  ],
  // ...
};
```

Now let's extend the runner function to calculate the score based on the budget:

**file-size plugin runnerFunction**

```typescript
// code-pushup.config.ts

// ...

async function runnerFunction(options: Options): Promise<AuditOutputs> {
  // ...

  // score audit based on budget
  const errorCount = data.map(({ size }) => options.budget < size).filter(Boolean).length;
  if (errorCount) {
    fileSizeAuditOutput: AuditOutput = {
      ...fileSizeAuditOutput,

>       // score is factor of overbudget files to total files

Reviewer: ```suggestion
      // score is percentage of over-budget files
```

This helps to make the plugin results more actionable and valuable for the user.

We will extend the `Options` type with a size `budget` property and see how we can use `details` in `AuditOutput` to
show which files exceed the defined budget.

**file-size plugin runnerFunction**

```typescript
// code-pushup.config.ts
// ...
import { formatBytes } from '@code-pushup/utils';

async function runnerFunction(options: Options): Promise<AuditOutputs> {
  // ...

  // assert file size information with budget
  const issues = data.map(({ file, size }) => assertFileSizeInfo(file, size, options.budget));

  // add details if issues given
  if (issues.length) {
    fileSizeAuditOutput = {
      ...fileSizeAuditOutput,
      details: {
        issues,
      },
    };
  }

  return [fileSizeAuditOutput];
}

// assert file size info with budget
export function assertFileSize(file: string, size: number, budget?: number): Issue {
  let severity: IssueSeverity = 'info';
  let message = `File ${basename(filePath)} OK`;

  if (budget !== undefined) {
    // set severity to error if budget exceeded
    if (budget < size) {
      severity = 'error';

>       message = `File ${basename(filePath)} is ${formatBytes(size - budget)} bytes too big. ( budget: ${formatBytes(budget)})`;

Reviewer: I'd specify `basename` import as well.

    }
  }

  return {
    message,
    severity,
    // add source attributes
    source: {
      file,
    },
  };
}
```

> Test the output by running `npx code-pushup collect --no-progress --format=md`.
> 
> The CLI argument `--format=md` will create an additional file containing our created detail information form above.
> 
> You should see a newly created file `report.md` crated in the folder `.code-pushup` in your current working directory.
> 
> The `report.md` file should contain a similar content like the following:

Reviewer 2: ```suggestion
Test the output by running `npx code-pushup collect --no-progress --format=md`.
The CLI argument `--format=md` will create an additional file containing our created detail information form above.
You should see a new file `report.md` created in the folder `.code-pushup` in your current working directory.
The `report.md` file should contain a similar content like the following:
```
I would not have the line breaks there.

**report.md**

```md
<!-- ... -->

### File Size Audit (File Size)

<details>
  <summary>🟨 <b>2 files</b> (score: 50)</summary>
  <h4>Issues</h4>
  <table>
    <tr>
      <th>Severity</th>
      <th>Message</th>
      <th>Source file</th>
      <th>Line(s)</th>
    </tr>
    <tr>
      <td>🚨 <i>error</i></td>
      <td>File file-a.js is 17.31 kB bytes too big. ( budget: 41.02 kB)</td>
      <td><code>/src/file-a.js</code></td>
      <td></td>
    </tr>
    <tr>
      <td>ℹ️ <i>info</i></td>
      <td>File file-b.js OK</td>
      <td><code>/src/file-b.js</code></td>
      <td></td>
    </tr>
  </table>
</details>
```

> ## Plugins and categories

Reviewer 2: An intro text is missing here. What is this section trying to convey?

**lighthouse plugin**

```typescript
// code-pushup.config.ts
import { CoreConfig } from '@code-pushup/core';

export default {
  // ...
  categories: [
    {
      slug: 'performance',
      title: 'Performance',
      description: "These encapsulate your app's current performance and opportunities to improve it.",
      refs: [
        {
          type: 'audit',
          slug: 'largest-contentful-paint',
          plugin: 'lightouse',
          weight: 1,
        },
      ],
    },
  ],
} satisfies CoreConfig;
```

Test the output by running `npx code-pushup collect --no-progress`.

**stdout of basic lighthouse plugin**

```sh
Code PushUp Report - @code-pushup/core@x.y.z

> Chrome Lighthosue audits

Reviewer: ```suggestion
Chrome Lighthouse audits
```

  This ensures we avoid overwrites of other existing files and is needed to be able to load the generated lighthouse
  report.
- To reduce the output you can execute only specific audits with the `--onlyAudits` option
  e.g.: `--onlyAudits=largest-contentful-paint`.  
  This will significantly reduce the time lighthouse takes to run.
- If we want to run the script in the background we can execute lighthouse headless with the
  flag `--chrome-flags="--headless=new"`.  
  It also is helpfun when executing lighthouse in th CI.

The basic implementation of a `RunnerConfig` for the above command looks like this:

```typescript
import { join } from 'path';
import { AuditOutputs } from '@code-pushup/models';

type Options = {
  url: string;
  onlyAudits: string;
};

function runnerConfig(options: Options): RunnerConfig {
  const { url, onlyAudits } = options;
  const outputFile = join(process.cwd(), '.code-pushup', 'lighthouse-report.json');
  return {
    // npx lighthouse https://example.com --output=json --outputFile=lighthouse-report.json  --onlyAudits=largest-contentful-paint
    command: 'npx',

>     args: objectToCliArgs({

Reviewer 3: Must the `objectToCliArgs` fn be imported from some of our packages?

      _: ['lighthouse', url],
      output: 'json',

>       ['output-path']: outputFile,

Reviewer 3: ```suggestion
      'output-path': outputFile,
```

      onlyAudits,
    }),
    outputFile,
    // implementation follows below
    outputTransform: async (output: string) => {
      // return dummy audit outputs. Correct implementation follows below
      return [
        {
          slug: lcpAudit.slug,
          score: 0,
          value: 0,
        },

>       ] as AuditOutputs;

Reviewer 3: Maybe better use `satisfies`?

```suggestion
      ] satisfies AuditOutputs;
```

    }),
  ],
  // ...
};

async function create(options: Options): Promise<PluginConfig> {
  // ...
  return {
    //...
    runner: await runnerConfig(options),
  };
}
```

Now we can execute the CLI with `npx code-pushup collect --no-progress` and see a similar output as the following:

```sh
Code PushUp Report - @code-pushup/core@x.y.z

Chrome Lighthosue audits
● Largest Contentful Paint                                                0

Categories
┌──────────┬───────┬────────┐
│ Category │ Score │ Audits │
└──────────┴───────┴────────┘

Made with ❤ by code-pushup.dev
```

#### implement a `outputTransform` function

As the result of the lighthouse run has a different shape than the required `AuditOutputs` we need to add
a `outputTransform` and implement the transform form a lighthouse report to audit outputs.

```typescript
import { Result } from 'lighthouse';
import { AuditOutputs } from './plugin-process-output';

function outputTransform(output: string): AuditOutputs {
  const lhr = JSON.parse(output as Result);
  return Object.values(lhr.audits).map(({ id: slug, score, numericValue: value, displayValue, description }) => ({
    slug,
    score: score || 0,

>     value: value ? parseInt(value.toString()) : 0,

Reviewer 3: Can we use parseInt with the `radix` parameter?

      - run: npx code-pushup autorun --upload.apiKey=${{ secrets.PORTAL_API_KEY }}
```

## Custom Plugins

> We provide comprehensive documentation on [how to create a custom plugin](@TODO).

Reviewer 2: Is this todo desired?

The repository also maintains a set of plugin examples showcasing different scenarios.  
Each example is fully tested to give demonstrate best practices for plugin testing.

**Example for custom plugins:**

> - [File Size](@TODO) -

Reviewer 2: Is this todo desired?
